From 0c1696770141b8ae308d90e5ac82abd3d93bf43a Mon Sep 17 00:00:00 2001
From: "Joel Fernandes (Google)" <joel@joelfernandes.org>
Date: Sat, 7 Jan 2023 04:09:04 +0000
Subject: [PATCH 10/20] ANDROID: rcu: Disable lazy if call_rcu() called when
 GPs expedited

During suspend, we see failures to suspend 1 in 300-500 suspends.
Looking closer, it appears that we are queuing lazy callbacks even
though rcu_gp_is_expedited(). These delays appear to not be very welcome
by the suspend/resume code as evidenced by these occasional suspend
failures.

This commit therefore checks if rcu_gp_is_expedited() and ignores the
lazy hint if so.

Ignoring the lazy hint if rcu_gp_is_expedited() makes the 3000
suspend/resume cycles pass reliably on a 12th gen 12-core Intel CPU.

[
  This is a temporary patch as it is a critical issue that we need to
  fix in product and the upstream fix will take some time.
  Once upstreamed, we can revert this and land the upstream one.
]

Bug: 258241771
BUG: 263554761
TEST=suspend_stress_test
(am from https://patchwork.kernel.org/patch/13094428/)
(also found at https://lore.kernel.org/r/20230109221756.1361168-1-joel@joelfernandes.org)

Fixes: 3cb278e73be5 ("rcu: Make call_rcu() lazy to save power")
Change-Id: I55430a6c6cc20f6e09dee45477c485d650e405c3
Signed-off-by: Joel Fernandes (Google) <joel@joelfernandes.org>
Reviewed-on: https://chromium-review.googlesource.com/c/chromiumos/third_party/kernel/+/4150549
Auto-Submit: Joel Fernandes <joelaf@google.com>
Tested-by: Joel Fernandes <joelaf@google.com>
Reviewed-by: Ross Zwisler <zwisler@google.com>
Commit-Queue: Joel Fernandes <joelaf@google.com>
(cherry picked from commit f8a532c737b798d7540f6b2b5a24bcc59f71264b)
Signed-off-by: Qais Yousef <qyousef@google.com>

Upstream: none. Upstream later fixed the same suspend failures with 6efdda8bec29 ("rcu: Track laziness during boot and suspend") and cf7066b97e27 ("rcu: Disable laziness if lazy-tracking says so"); not taken, since in 5.10 rcu_gp_is_expedited() is already true during in-kernel boot (rcu_expedited_nesting starts at 1) and across suspend/hibernation (rcu_pm_notify() calls rcu_expedite_gp()).
ACK android13-5.10: 9788d1e389956c289cd43d776ec24b2b63473034, taken from arter97/master
---
 kernel/rcu/tree.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/kernel/rcu/tree.c b/kernel/rcu/tree.c
index bf3d26052fa0..afdca180d490 100644
--- a/kernel/rcu/tree.c
+++ b/kernel/rcu/tree.c
@@ -2949,11 +2949,11 @@ static void check_cb_ovld(struct rcu_data *rdp)
 }
 
 static void
-__call_rcu_common(struct rcu_head *head, rcu_callback_t func, bool lazy)
+__call_rcu_common(struct rcu_head *head, rcu_callback_t func, bool lazy_in)
 {
 	unsigned long flags;
 	struct rcu_data *rdp;
-	bool was_alldone;
+	bool was_alldone, lazy;
 
 	/* Misaligned rcu_head! */
 	WARN_ON_ONCE((unsigned long)head & (sizeof(void *) - 1));
@@ -2978,6 +2978,7 @@ __call_rcu_common(struct rcu_head *head, rcu_callback_t func, bool lazy)
 	local_irq_save(flags);
 	kasan_record_aux_stack(head);
 	rdp = this_cpu_ptr(&rcu_data);
+	lazy = lazy_in && !rcu_gp_is_expedited();
 
 	/* Add the callback to our list. */
 	if (unlikely(!rcu_segcblist_is_enabled(&rdp->cblist))) {
